### PR TITLE
First working concept of the Terms Stats Plus

### PR DESCRIPTION
**Pull request by sadasant**
Currently only filters inline and has the "Add as a filter"
functionality.

Feedback is appreciated!

### CODE REVIEW

**Comment by decrapifier** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Your PR has lint errors. Please fix these and commit them.
      
<details>
  <summary>ResultTable.js</summary>
  <p><a href="https://github.com/smartprocure/contexture-react/blob/feature/TermsStatsPlus/src/exampleTypes/ResultTable.js" target="_blank">src/exampleTypes/ResultTable.js</a></p>
  
  <strong>'flattenPlainObject' is defined but never used.</strong>
  <code>Line 11: import { flattenPlainObject } from '../utils/futil'</code>

  <strong>'../utils/schema' import is duplicated.</strong>
  <code>Line 12: import { getRecord, getResults, inferSchema } from '../utils/schema'</code>
</details>
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  The changelog has not been updated. Please update the changelog.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  The version was not updated. Please update the version.
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  The README has not been updated. Please update the README.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  Could not run `duti:fix` command successfully
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  Could not find any browser results.
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: src/exampleTypes/TermsStatsTable.js [hunk 1]
```diff
@@ -1,12 +1,83 @@
...
+    <div>
+      <div>
+        Filter:
+        <input
+          type="text"
+          value={node.filter}
+          onChange={async e => {
+            await tree.mutate(node.path, { filter: e.target.value })
+            // TODO: This seems like a contexture-client issue. We shouldn't need this line:
+            await tree.refresh(node.path)
```
Comment (daedalus28) [inline src/exampleTypes/TermsStatsTable.js]: It is a missing reactor: https://github.com/smartprocure/contexture-client/blob/master/src/exampleTypes.js#L186

Just add a self reactor for `filter` there

File: src/exampleTypes/TermsStatsTable.js [hunk 2]
```diff
@@ -1,12 +1,83 @@
...
+            <Column
+              label="Controls"
+              expand={{
+                display: (value, record) => (
+                  <div>
+                    <button
+                      onClick={async () => {
+                        // TODO: Could this be smaller?
+                        let schema = _.flow(
+                          _.merge(infer && inferSchema(node)),
```
Comment (daedalus28) [inline src/exampleTypes/TermsStatsTable.js]: Why would you infer the schema on a terms stats table? Key, count, etc won't be in the schema since they're aggregates

File: src/exampleTypes/TermsStatsTable.js [hunk 3]
```diff
@@ -1,12 +1,83 @@
...
+                          applyDefaults,
+                          _.values,
+                          _.orderBy('order', 'desc')
```
Comment (daedalus28) [inline src/exampleTypes/TermsStatsTable.js]: Why would order matter?

File: src/exampleTypes/TermsStatsTable.js [hunk 4]
```diff
@@ -1,12 +1,83 @@
...
+                        let field = node.key_field
+                        let filter =
+                          criteria &&
+                          _.find({ field }, tree.getNode(criteria).children)
+
+                        if (!filter) {
+                          await tree.add(criteria, {
+                            key: _.uniqueId('add'),
+                            field,
+                            type: _.find({ field }, schema).typeDefault,
```
Comment (daedalus28) [inline src/exampleTypes/TermsStatsTable.js]: In this case, it pretty much has to be a `facet` because you're mutating `values`

File: src/exampleTypes/TermsStatsTable.js [hunk 5]
```diff
@@ -1,12 +1,83 @@
...
+                          })
+                          filter = _.find(
+                            { field },
+                            tree.getNode(criteria).children
+                          )
+                        }
+
+                        await tree.mutate(filter.path, {
+                          values: [
+                            ..._.castArray(filter.values.slice()),
```
Comment (daedalus28) [inline src/exampleTypes/TermsStatsTable.js]: pretty sure we want to filter to _only_ `record.key`. If we just added it to the list, the results wouldn't change because the row already matches in the search or it wouldn't be there. You probably also want to ensure it's mode `include`

File: src/exampleTypes/TermsStatsTable.js [hunk 6]
```diff
@@ -1,12 +1,83 @@
...
+                            record.key,
+                          ],
+                        })
+                      }}
+                    >
+                      Add as Filter
+                    </button>
+                    <button>Open Profile (TODO)</button>
```
Comment (daedalus28) [inline src/exampleTypes/TermsStatsTable.js]: This should be an additional column which can live in user land in the app repo because it's pure business logic.

File: src/exampleTypes/TermsStatsTable.js (hunk 4 above)
Comment (daedalus28) [reply src/exampleTypes/TermsStatsTable.js]: in fact, since you don't need schema here for type default, I don't think you need to include schema stuff in here at all

File: stories/imdb/stories/filterList.js [hunk 7]
```diff
@@ -136,6 +137,9 @@ export default () => (
                   format={formatYear}
                 />
                 <TermsStatsTable
+                  infer
+                  fields={_.pick(['genres'], schemas.movies.fields)}
```
Comment (daedalus28) [inline stories/imdb/stories/filterList.js]: why is this needed?

File: stories/imdb/stories/filterList.js [hunk 8]
```diff
@@ -136,6 +137,9 @@ export default () => (
                   format={formatYear}
                 />
                 <TermsStatsTable
+                  infer
```
Comment (daedalus28) [inline stories/imdb/stories/filterList.js]: this doesn't make sense

File: stories/imdb/stories/filterList.js [hunk 9]
```diff
@@ -67,6 +67,7 @@ let tree = Contexture({
       type: 'terms_stats',
       key_field: 'genres',
       value_field: 'metaScore',
+      filter: '',
```
Comment (daedalus28) [inline stories/imdb/stories/filterList.js]: this should be solved with a `default` in contexture-client

File: src/exampleTypes/TermsStatsTable.js (hunk 1 above)
Comment (daedalus28) [reply src/exampleTypes/TermsStatsTable.js]: you can also use tree.lens + futil.domLens.value:

```js
<input
  type="text"
  {...F.domLens.value(tree.lens(node.path, 'filter'))}
/>```

File: src/exampleTypes/TermsStatsTable.js [hunk 10]
```diff
@@ -1,12 +1,83 @@
...
+import { applyDefaults, inferSchema } from '../utils/schema'
 
 let TermsStatsTable = injectTreeNode(
-  observer(({ node, ...props }) => (
-    <ExpandableTable {...props} data={node.context.terms} />
+  observer(({ infer, fields, node, criteria, tree, children, ...props }) => (
+    <div>
+      <div>
+        Filter:
+        <input
```
Comment (daedalus28) [inline src/exampleTypes/TermsStatsTable.js]: This should take Input as a prop and be passed in example types

File: src/exampleTypes/TermsStatsTable.js [hunk 11]
```diff
@@ -1,13 +1,80 @@
...
+      MoreControls = 'div',
+      Input = 'input',
+      ...props
+    }) => (
+      <div>
+        <div>
+          Filter:
+          <Input
+            type="text"
+            value={node.filter}
```
Comment (daedalus28) [inline src/exampleTypes/TermsStatsTable.js]: you don't need `value` here if you're using `domLens.value`

File: src/exampleTypes/TermsStatsTable.js [hunk 12]
```diff
@@ -1,13 +1,80 @@
...
+          {...{
+            ...props,
+            children: [
+              ...children,
+              <Column
+                label="Controls"
+                expand={{
+                  display: (value, record) => (
+                    <div>
+                      <button
```
Comment (daedalus28) [inline src/exampleTypes/TermsStatsTable.js]: We should only do this if criteria is provided - if there is no criteria, this will break (and criteria should be optional)
